components/raftstore-v2/src/operation/query/mod.rs

```diff
@@ -2,41 +2,318 @@
+                    let read_resp = ReadResponse::new(0);
+                    ch.set_result(QueryResult::Read(read_resp));
+                }
+                _ => {
+                    unimplemented!();
+                }
+            };
         } else {
             self.fsm.peer_mut().on_query_status(&req, ch);
         }
     }
 }
 
 impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
+    fn validate_query_msg(
+        &mut self,
+        msg: &RaftCmdRequest,
+        raft_metrics: &mut RaftMetrics,
+    ) -> Result<()> {
+        // check query specific requirements
+        if msg.has_admin_request() {
+            return Err(box_err!("PeerMsg::RaftQuery does not allow admin requests"));
+        }
+
+        // check query specific requirements
+        for r in msg.get_requests() {
+            if r.get_cmd_type() != CmdType::Get
+                && r.get_cmd_type() != CmdType::Snap
+                && r.get_cmd_type() != CmdType::ReadIndex
+            {
+                return Err(box_err!(
+                    "PeerMsg::RaftQuery does not allow write requests: {:?}",
+                    r.get_cmd_type()
+                ));
+            }
+        }
+
+        // Check store_id, make sure that the msg is dispatched to the right place.
+        if let Err(e) = util::check_store_id(msg, self.peer().get_store_id()) {
+            raft_metrics.invalid_proposal.mismatch_store_id.inc();
+            return Err(e);
+        }
+
+        let flags = WriteBatchFlags::from_bits_check(msg.get_header().get_flags());
+        if flags.contains(WriteBatchFlags::STALE_READ) {
+            return Err(box_err!(
+                "PeerMsg::RaftQuery should not get stale read requests"
+            ));
+        }
+
+        // TODO: add flashback_state check
+
+        // Check whether the store has the right peer to handle the request.
+        let leader_id = self.leader_id();
+        let request = msg.get_requests();
+
+        // TODO: add force leader
+
+        // ReadIndex can be processed on the replicas.
+        let is_read_index_request =
+            request.len() == 1 && request[0].get_cmd_type() == CmdType::ReadIndex;
+
+        let allow_replica_read = msg.get_header().get_replica_read();
+        let flags = WriteBatchFlags::from_bits_check(msg.get_header().get_flags());
+        let allow_stale_read = flags.contains(WriteBatchFlags::STALE_READ);
+        if !self.is_leader() && !is_read_index_request && !allow_replica_read && !allow_stale_read {
+            raft_metrics.invalid_proposal.not_leader.inc();
+            return Err(Error::NotLeader(self.region_id(), None));
+        }
+
+        // peer_id must be the same as peer's.
+        if let Err(e) = util::check_peer_id(msg, self.peer_id()) {
+            raft_metrics.invalid_proposal.mismatch_peer_id.inc();
+            return Err(e);
+        }
+        // check whether the peer is initialized.
+        if !self.storage().is_initialized() {
```

Comment:
I don't think we need to check this on v2. If it's lease read, then it can't be uninitialized; if it's a follower read, then it must wait for apply, so it will be initialized in the end.

Reply:
So, why v1 need this check?

Reply:
> I don't think we need to check this on v2. If it's lease read, then it can't be uninitialized; if it's a follower read, then it must wait for apply, so it will be initialized in the end.

So we should add assert instead then.  Also why it needs the check in v1?

Reply:
So in v1 it's mainly for replica read scenario, which is not needed in v2's raftstore. 
https://github.com/tikv/tikv/issues/5875